### PR TITLE
Have `get_route` and `Route` take `RouteParameters`

### PR DESCRIPTION
**Pull request by tnull**
These are the first few commits of #2417 split out to it's own PR in order to avoid having too many rebase conflicts.

### CODE REVIEW

File: lightning/src/routing/router.rs [hunk 1]
```diff
@@ -435,6 +435,13 @@ pub struct RouteParameters {
 	pub final_value_msat: u64,
 }
 
+impl RouteParameters {
+	/// Constructs [`RouteParameters`] from the given [`PaymentParameters`] and a payment amount.
+	pub fn from_payment_params_and_value(payment_params: PaymentParameters, final_value_msat: u64) -> Self {
```
Comment (valentinewallace) [inline lightning/src/routing/router.rs:462]: I think this could be called `fn new` but fine as is

File: lightning/src/routing/router.rs [hunk 2]
```diff
@@ -337,10 +337,10 @@ pub struct Route {
 	/// [`BlindedTail`]s are present, then the pubkey of the last [`RouteHop`] in each path must be
 	/// the same.
 	pub paths: Vec<Path>,
-	/// The `payment_params` parameter passed via [`RouteParameters`] to [`find_route`].
+	/// The `route_params` parameter passed [`find_route`].
```
Comment (valentinewallace) [inline lightning/src/routing/router.rs:340]: nit: s/passed/passed to

File: lightning/src/routing/router.rs [hunk 3]
```diff
@@ -383,8 +383,11 @@ impl Writeable for Route {
 			}
 		}
 		write_tlv_fields!(writer, {
-			(1, self.payment_params, option),
+			// For compatibility with LDK versions prior to 0.0.117, we the individual
```
Comment (valentinewallace) [inline lightning/src/routing/router.rs:395]: nit: `s/we the/we take the` ?

File: lightning/src/routing/router.rs [hunk 4]
```diff
@@ -419,14 +423,23 @@ impl Readable for Route {
 				path.blinded_tail = blinded_tail_opt;
 			}
 		}
-		Ok(Route { paths, payment_params })
+
+		// If we previously wrote, the corresponding fields, reconstruct RouteParameters.
```
Comment (valentinewallace) [inline lightning/src/routing/router.rs:436]: nit: s/wrote,/wrote

**Review by valentinewallace — APPROVED** [review]
Nothing blocking, thanks for undertaking this annoying sidequest 😛

File: lightning/src/routing/router.rs [hunk 5]
```diff
@@ -357,8 +357,12 @@ impl Route {
 		overpaid_value_msat + self.paths.iter().map(|path| path.fee_msat()).sum::<u64>()
 	}
 
-	/// Returns the total amount paid on this [`Route`], excluding the fees. Might be more than
-	/// requested if we had to reach htlc_minimum_msat.
+	/// Returns the total amount paid on this [`Route`], excluding the fees.
+	///
+	/// Might be more than requested as part of the given [`RouteParameters::final_value_msat`] if
+	/// we had to reach the [`htlc_minimum_msat`] limit.
```
Comment (jkczyz) [inline lightning/src/routing/router.rs:363]: nit: "limits" since it is per hop

**Review by jkczyz — DISMISSED** [review]
Also nothing worth blocking.

**Review by valentinewallace — APPROVED** [review]

**Review by jkczyz — APPROVED** [review]

**Pull request merged**